cmd/helm/completion.go

```diff
@@ -145,148 +145,25 @@ fi
-		shift
-		# fake Bash 4 to disable "complete -o nospace". Instead
-		# "compopt +-o nospace" is used in the code to toggle trailing
-		# spaces. We don't support that, but leave trailing spaces on
-		# all the time
-		if [ "$1" = "__helm_compopt" ]; then
-			echo builtin
-			return 0
-		fi
-	fi
-	type "$@"
-}
-__helm_compgen() {
-	local completions w
-	completions=( $(compgen "$@") ) || return $?
-	# filter by given word as prefix
-	while [[ "$1" = -* && "$1" != -- ]]; do
-		shift
-		shift
-	done
-	if [[ "$1" == -- ]]; then
-		shift
-	fi
-	for w in "${completions[@]}"; do
-		if [[ "${w}" = "$1"* ]]; then
-			# Use printf instead of echo because it is possible that
-			# the value to print is -n, which would be interpreted
-			# as a flag to echo
-			printf "%s\n" "${w}"
-		fi
-	done
-}
-__helm_compopt() {
-	true # don't do anything. Not supported by bashcompinit in zsh
-}
-__helm_ltrim_colon_completions()
-{
-	if [[ "$1" == *:* && "$COMP_WORDBREAKS" == *:* ]]; then
-		# Remove colon-word prefix from COMPREPLY items
-		local colon_word=${1%${1##*:}}
-		local i=${#COMPREPLY[*]}
-		while [[ $((--i)) -ge 0 ]]; do
-			COMPREPLY[$i]=${COMPREPLY[$i]#"$colon_word"}
-		done
-	fi
-}
-__helm_get_comp_words_by_ref() {
-	cur="${COMP_WORDS[COMP_CWORD]}"
-	prev="${COMP_WORDS[${COMP_CWORD}-1]}"
-	words=("${COMP_WORDS[@]}")
-	cword=("${COMP_CWORD[@]}")
-}
-__helm_filedir() {
-	local RET OLD_IFS w qw
-	__debug "_filedir $@ cur=$cur"
-	if [[ "$1" = \~* ]]; then
-		# somehow does not work. Maybe, zsh does not call this at all
-		eval echo "$1"
-		return 0
-	fi
-	OLD_IFS="$IFS"
-	IFS=$'\n'
-	if [ "$1" = "-d" ]; then
-		shift
-		RET=( $(compgen -d) )
-	else
-		RET=( $(compgen -f) )
-	fi
-	IFS="$OLD_IFS"
-	IFS="," __debug "RET=${RET[@]} len=${#RET[@]}"
-	for w in ${RET[@]}; do
-		if [[ ! "${w}" = "${cur}"* ]]; then
-			continue
-		fi
-		if eval "[[ \"\${w}\" = *.$1 || -d \"\${w}\" ]]"; then
-			qw="$(__helm_quote "${w}")"
-			if [ -d "${w}" ]; then
-				COMPREPLY+=("${qw}/")
-			else
-				COMPREPLY+=("${qw}")
-			fi
-		fi
-	done
-}
-__helm_quote() {
-	if [[ $1 == \'* || $1 == \"* ]]; then
-		# Leave out first character
-		printf %q "${1:1}"
-	else
-		printf %q "$1"
-	fi
-}
-autoload -U +X bashcompinit && bashcompinit
-# use word boundary patterns for BSD or GNU sed
-LWORD='[[:<:]]'
-RWORD='[[:>:]]'
-if sed --help 2>&1 | grep -q 'GNU\|BusyBox'; then
-	LWORD='\<'
-	RWORD='\>'
+	err := cmd.Root().GenZshCompletion(out)
+
+	// In case the user renamed the helm binary (e.g., to be able to run
+	// both helm2 and helm3), we hook the new binary name to the completion function
+	if binary := filepath.Base(os.Args[0]); binary != "helm" {
+		renamedBinaryHook := `
+# Hook the command used to generate the completion script
+# to the helm completion function to handle the case where
+# the user renamed the helm binary
+if [[ $(type -t compopt) = "builtin" ]]; then
+    complete -o default -F __start_helm %[1]s
+else
+    complete -o default -o nospace -F __start_helm %[1]s
```

Comment:
This whole if statement can be removed since it's bash syntax.

It should be replaced by the zsh version which is simply: `compdef _helm helm`

Cobra does not add that line at all because it favors installing the script on `fpath` instead of sourcing it.  To be backwards compatible though, helm should support sourcing the script. Therefore you'll need to add that line at the very end of the script generated by Cobra all the time.

You would also need to add it in case the binary name is not `helm`.  We may also need to change the first line of the generated script to the different binary name, to support putting the script on `fpath`; we'd have to test it to be sure.

Reply:
I tested for sourcing for both cases using helm binary and different name binary it works. Also `fpath` for helm binary works. But I am not sure how to check for changed name for `fpath`